scripts/peer/peer.py

```diff
@@ -0,0 +1,273 @@
+    # NB: these are the ones in lowercase (e.g., 'pc1')
+    covariates = pd.read_csv(covariates_path, sep='\t')
+    covariates = pd.merge(
+        scores, covariates, how='left', left_on='OneK1K_ID', right_on='sampleid'
+    ).drop(['pc1', 'pc2', 'pc3', 'pc4'], axis=1)
+    # check whether there's any missing data in the covariate data
+    print(covariates[covariates.sampleid.isna()])
+    # OneK1K ID 26_26 does not have any covariate data. Remove this sample and drop
+    # the OneK1K_ID, since it's redundant with sampleid
+    covariates = covariates[~covariates.sampleid.isna()].drop(
+        ['OneK1K_ID'], axis=1
+    )
+    # Match expression data to covariate data, since PEER needs expr and covs
+    # dfs to be the same length
+    expression = pd.read_csv(expression_file, sep='\t')
+    merged_expr_covs = pd.merge(
+        expression, covariates, how='right', left_on='sampleid', right_on='sampleid'
+    )
+    print(merged_expr_covs.shape[0] - merged_expr_covs.dropna().shape[0])
+    # 15 samples with no expression data, but do have covariate and genotype data
+    # This differs per celltype
+    # remove any rows with NA values
+    merged_expr_covs = merged_expr_covs.dropna()
+    expression = merged_expr_covs.drop(
+        ['PC1', 'PC2', 'PC3', 'PC4', 'sex', 'age'], axis=1
+    )
+    covariates = merged_expr_covs[
+        ['sampleid', 'PC1', 'PC2', 'PC3', 'PC4', 'sex', 'age']
+    ]
+    # make sure samples in covariates and expression data are equal
+    expression.sampleid.equals(covariates.sampleid)
+    # True
+    # remove sampleid from both covariates and expression dfs
+    expression.drop('sampleid', axis=1, inplace=True)
+    covariates.drop('sampleid', axis=1, inplace=True)
+    # finally, make sure age and sex in covariate df are int
+    covariates[['sex', 'age']] = covariates[['sex', 'age']].astype(int)
+
+    # return expression data and covariates
+    return covariates.to_csv(index=False), expression.to_csv(index=False)  # type: ignore
+
+
+def get_at_index(obj, idx):
+    """
+    get object index
+    """
+
+    return obj[idx]
+
+
+def run_peer_job(job: hb.batch.job.Job, expression_file, covariates_file):
+    """
+    Run peer analysis, except because peer is in Python2, we can't take
+    advantage of the python_job concept in Batch. Instead, we'll rely on
+    the two inputs `expression_file` and `covariates_file` to be TSV files.
+    """
+
+    job.image(PEER_DOCKER)
+
+    job.command(f'echo "expressions file:" && head {expression_file}')
+    job.command(f'echo "covariates file:" && head {covariates_file}')
+
+    # write python script to container
+    job.command(
+        """
+cat <<EOT >> run_peer.py
+
+import peer
+import numpy as np
+import pandas as pd
+
+def run_peer(expression_file, covariates_file, factors_output_path, weights_output_path, precision_output_path, residuals_output_path):
+    \"""
+    Get covariate data for each cell type
+    \"""
+
+    print 'Loading data'
```

Comment:
Should remove these debug statements, or make them more descriptive

Reply:
I removed them